src/libstd/ffi/c_str.rs

```diff
@@ -200,6 +200,38 @@ impl CString {
         CString { inner: v }
     }
 
+    /// Retakes ownership of a CString that was transferred to C.
+    ///
+    /// The only appropriate argument is a pointer obtained by calling
+    /// `into_raw`.
+    #[unstable(feature = "c_str_memory", reason = "recently added")]
+    pub unsafe fn from_raw(ptr: *const libc::c_char) -> CString {
+        let len = libc::strlen(ptr);
+        let len_with_nul = len as usize + 1;
+        let vec = Vec::from_raw_parts(ptr as *mut u8, len_with_nul, len_with_nul);
+        CString::from_vec_unchecked(vec)
+    }
+
+    /// Transfers ownership of the string to a C caller.
+    ///
+    /// The pointer must be returned to Rust and reconstituted using
+    /// `from_raw` to be properly deallocated. Specifically, one
+    /// should *not* use the standard C `free` function to deallocate
+    /// this string.
+    ///
+    /// Failure to call `from_raw` will lead to a memory leak.
+    #[unstable(feature = "c_str_memory", reason = "recently added")]
+    pub fn into_raw(self) -> *const libc::c_char {
+        // Resize the vector to fit - we need the capacity to be
+        // determinable from the string length, and shrinking to fit
+        // is the only way to be sure.
+        let mut vec = self.inner;
+        vec.shrink_to_fit();
```

Comment:
This is not a problem local to `CString` as we already depend on `shrink_to_fit` to actually "shrink to fit" for the [`Vec::into_boxed_slice`](https://github.com/rust-lang/rust/blob/9c303944a7c6767cd5cb708d25ffa27851fa994c/src/libcollections/vec.rs#L402-L415) method. 

@shepmaster could this actually switch to using `into_boxed_slice` actually? That way it pushes the responsibility onto `Vec<T>` and then it's guaranteed that the value coming out is OK to be freed. I think that this may involve changing `CString` to internally store `Box<[u8]>`, but that should be fine.